### PR TITLE
AP_Arming: Check all channels pass (<= min trim max)

### PR DESCRIPTION
**Pull request by WickedShell**
This sanity checks all servo and RC channels that the condition min <= trim <= max passes. Really funky things happen when this isn't true. (Note that min == trim == max is accepted as it's common to get this on unused channels during a RC cal with some GCS's).

### CODE REVIEW

File: libraries/AP_Arming/AP_Arming.cpp [hunk 1]
```diff
@@ -455,7 +456,27 @@ bool AP_Arming::hardware_safety_check(bool report)
 
 bool AP_Arming::rc_calibration_checks(bool report)
 {
-    return true;
+    bool check_passed = true;
+    for (uint8_t i = 0; i < NUM_RC_CHANNELS; i++) {
+        RC_Channel *ch = RC_Channels::rc_channel(i);
+        if (ch != nullptr) {
```
Comment (peterbarker) [inline libraries/AP_Arming/AP_Arming.cpp]: Early-`continue` instead to avoid indentation

File: libraries/AP_Arming/AP_Arming.cpp [hunk 2]
```diff
@@ -455,7 +456,27 @@ bool AP_Arming::hardware_safety_check(bool report)
 
 bool AP_Arming::rc_calibration_checks(bool report)
 {
-    return true;
+    bool check_passed = true;
+    for (uint8_t i = 0; i < NUM_RC_CHANNELS; i++) {
+        RC_Channel *ch = RC_Channels::rc_channel(i);
```
Comment (peterbarker) [inline libraries/AP_Arming/AP_Arming.cpp]: `const`

File: libraries/AP_Arming/AP_Arming.cpp [hunk 3]
```diff
@@ -455,7 +456,27 @@ bool AP_Arming::hardware_safety_check(bool report)
 
 bool AP_Arming::rc_calibration_checks(bool report)
```
Comment (peterbarker) [inline libraries/AP_Arming/AP_Arming.cpp]: Should be `const`; pre-existing issue :-)

File: libraries/AP_Arming/AP_Arming.cpp [hunk 4]
```diff
@@ -455,7 +456,27 @@ bool AP_Arming::hardware_safety_check(bool report)
...
 {
-    return true;
+    bool check_passed = true;
+    for (uint8_t i = 0; i < NUM_RC_CHANNELS; i++) {
+        RC_Channel *ch = RC_Channels::rc_channel(i);
+        if (ch != nullptr) {
+            uint16_t trim = ch->get_radio_trim();
```
Comment (peterbarker) [inline libraries/AP_Arming/AP_Arming.cpp]: `const`

File: libraries/AP_Arming/AP_Arming.cpp [hunk 5]
```diff
@@ -478,6 +499,31 @@ bool AP_Arming::manual_transmitter_checks(bool report)
     return true;
 }
 
+bool AP_Arming::servo_checks(bool report)
```
Comment (peterbarker) [inline libraries/AP_Arming/AP_Arming.cpp]: `const`

File: libraries/AP_Arming/AP_Arming.cpp [hunk 6]
```diff
@@ -478,6 +499,31 @@ bool AP_Arming::manual_transmitter_checks(bool report)
...
+{
+    bool check_passed = true;
+    for (uint8_t i = 0; i < NUM_SERVO_CHANNELS; i++) {
+        SRV_Channel *ch = SRV_Channels::srv_channel(i);
```
Comment (peterbarker) [inline libraries/AP_Arming/AP_Arming.cpp]: `const`

File: libraries/AP_Arming/AP_Arming.cpp [hunk 7]
```diff
@@ -478,6 +499,31 @@ bool AP_Arming::manual_transmitter_checks(bool report)
...
+        if (ch != nullptr && ch->get_function() != SRV_Channel::k_none) {
```
Comment (peterbarker) [inline libraries/AP_Arming/AP_Arming.cpp]: Early-`continue`?

I'd also break the thing up into two separate if-clauses, but that's more a personal-preference-thing that seems to be indulged...

File: libraries/AP_Arming/AP_Arming.cpp [hunk 8]
```diff
@@ -478,6 +499,31 @@ bool AP_Arming::manual_transmitter_checks(bool report)
...
+            uint16_t trim = ch->get_trim();
```
Comment (peterbarker) [inline libraries/AP_Arming/AP_Arming.cpp]: `const`

File: libraries/AP_Arming/AP_Arming.cpp (hunk 3 above)
Comment (WickedShell) [reply libraries/AP_Arming/AP_Arming.cpp]: I let this slip through, on account of needing to then touch copter + sub...

**Comment by WickedShell** [comment]
That's part of why I accept `(= min trim max)` as valid, as that should vastly minimize the number of unused channel failures. Servos we know what the function is, which made it easy to sanity check. On the RC side I didn't see a good approach to that, so I simply checked all the channels.

File: libraries/AP_Arming/AP_Arming.cpp [hunk 9]
```diff
@@ -455,7 +455,28 @@ bool AP_Arming::hardware_safety_check(bool report)
 
 bool AP_Arming::rc_calibration_checks(bool report)
 {
-    return true;
+    bool check_passed = true;
+    for (uint8_t i = 0; i < NUM_RC_CHANNELS; i++) {
+        const RC_Channel *ch = RC_Channels::rc_channel(i);
+        if (ch != nullptr) {
```
Comment (OXINARF) [inline libraries/AP_Arming/AP_Arming.cpp]: `ch == nullptr`

File: libraries/AP_Arming/AP_Arming.cpp [hunk 10]
```diff
@@ -499,6 +500,33 @@ bool AP_Arming::manual_transmitter_checks(bool report)
     return true;
 }
 
+bool AP_Arming::servo_checks(bool report) const
+{
+    bool check_passed = true;
+    for (uint8_t i = 0; i < NUM_SERVO_CHANNELS; i++) {
+        const SRV_Channel *ch = SRV_Channels::srv_channel(i);
+        if (ch != nullptr && ch->get_function() != SRV_Channel::k_none) {
```
Comment (OXINARF) [inline libraries/AP_Arming/AP_Arming.cpp]: `ch == nullptr || ch->get_function() == SRV_Channel::k_none`

**Comment by WickedShell** [comment]
@OXINARF amazing how that is...

**Comment by OXINARF** [comment]
Merged, thanks!

**Pull request merged**